src/refract/Element.h

```diff
@@ -0,0 +1,444 @@
+            virtual ~MemberElementCollection();
+        };
+
+        MemberElementCollection meta;
+        MemberElementCollection attributes;
+
+        /**
+         * return "name" of element
+         * usualy injected by "trait", but you can set own
+         * via pair method `element(std::string)`
+         */
+        virtual std::string element() const = 0;
+        virtual void element(const std::string&) = 0;
+
+        // NOTE: Visiting is now handled by inheritance from `VisitableBy`
+        // it uses internally C++ RTTI via `dynamic_cast<>`. 
+        // And accepts all visitors declared in typelist IElement::Visitors
+        //
+        // Alternative solution to avoid RTTI:
+        // Add overrided virtual function `content` for every one type of `Visitor`
+        
+        // NOTE: probably rename to Accept
+        virtual void content(IVisitor& v) const = 0;
+
+        virtual bool empty() const
+        {
+            return !hasContent;
+        }
+
+        virtual bool compactContent() const
+        {
+            return useCompactContent;
+        }
+
+        virtual void renderCompactContent(bool compact)
+        {
+            useCompactContent = compact;
+        }
+
+        virtual ~IElement()
+        {
+        }
+    };
+
+    struct IVisitor
+    {
+        virtual ~IVisitor(){};
+    };
+
+    struct ComparableVisitor : IVisitor
+    {
+        std::string compare_to;
+        bool result;
+
+        ComparableVisitor(const std::string& str) : compare_to(str), result(false)
+        {
+        }
+
+        template <typename T, typename U>
+        bool IsEqual(const T& first, const U& second)
+        {
+            return false;
+        }
+
+        template <typename T>
+        bool IsEqual(const T& first, const T& second)
+        {
+            return first == second;
+        }
+
+        template <typename E>
+        void visit(const E& e)
+        {
+            result = IsEqual(compare_to, e.value);
+        }
+
+        virtual void visit(const IElement& e)
+        {
+            throw LogicError("Fallback impl - behavioration for Base class IElement is not defined");
+        }
+
+        operator bool() const
+        {
+            return result;
+        }
+    };
+
+    struct SerializeVisitor : IVisitor
+    {
+
+        sos::Object result;
+        sos::Base partial;
+        std::string key;
+
+        SerializeVisitor() : partial(sos::Null())
+        {
+        }
+
+        void visit(const IElement& e);
+        void visit(const NullElement& e);
+        void visit(const StringElement& e);
+        void visit(const NumberElement& e);
+        void visit(const BooleanElement& e);
+        void visit(const ArrayElement& e);
+        void visit(const MemberElement& e);
+        void visit(const ObjectElement& e);
+
+        sos::Object get()
+        {
+            return result;
+        }
+    };
+
+    struct SerializeCompactVisitor : IVisitor
```

Comment:
I am just wondering if all these serialize stuff should be in `Element.h` itself. Can we use another header file?